```console
$ cat qnx_build.sh
#!/bin/bash
source ~/qnx700/qnxsdp-env.sh
export build_env='
         CC_aarch64-unknown-nto-qnx700=qcc
        CXX_aarch64-unknown-nto-qnx700=qcc
     CFLAGS_aarch64-unknown-nto-qnx710=-Vgcc_ntoaarch64le_cxx
         AR_aarch64_unknown_nto_qnx700=ntoaarch64-ar
         CC_x86_64-pc-nto-qnx700=qcc
        CXX_x86_64-pc-nto-qnx700=qcc
     CFLAGS_x86_64-pc-nto-qnx710=-Vgcc_ntox86_64_cxx
         AR_x86_64_pc_nto_qnx700=ntox86_64-ar
         CC_i586-pc-nto-qnx700=qcc
        CXX_i586-pc-nto-qnx700=qcc
     CFLAGS_i586-pc-nto-qnx710=-Vgcc_ntox86_cxx
         AR_i586_pc_nto_qnx700=ntox86-ar
'
env $build_env ./x.py build \
  --target aarch64-apple-darwin,aarch64-unknown-nto-qnx700 \
  rustc library/core library/alloc library/std src/tools/remote-test-server
$ ./qnx_build.sh

< time passes >

Building stage1 library artifacts {alloc, core, std} (aarch64-apple-darwin -> aarch64-unknown-nto-qnx700)
   Compiling std v0.0.0 (/Users/jonathan/Documents/open-source/rust/library/std)
error: linking with `qcc` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1/lib/rustlib/aarch64-apple-darwin/bin:/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1/lib/rustlib/aarch64-apple-darwin/bin:/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1/lib/rustlib/aarch64-apple-darwin/bin:/Users/jonathan/qnx700/host/darwin/x86_64/usr/bin:/Users/jonathan/.qnx/bin:/Users/jonathan/qnx700/jre/bin:/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/opt/X11/bin:/Library/Apple/usr/bin:/Library/TeX/texbin:/Users/jonathan/.cargo/bin:/Applications/iTerm.app/Contents/Resources/utilities:/Users/jonathan/.local/bin" VSLANG="1033" "qcc" "-Wl,--version-script=/var/folders/0r/8h59vybs29n7zzrgx_ykfgpw0000gn/T/rustc64EQxG/list" "-Wl,--no-undefined-version" "-Vgcc_ntoaarch64le_cxx" "/var/folders/0r/8h59vybs29n7zzrgx_ykfgpw0000gn/T/rustc64EQxG/symbols.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.std.b489edb20527c47c-cgu.00.rcgu.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.std.b489edb20527c47c-cgu.01.rcgu.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.std.b489edb20527c47c-cgu.02.rcgu.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.std.b489edb20527c47c-cgu.03.rcgu.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.std.b489edb20527c47c-cgu.04.rcgu.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.std.b489edb20527c47c-cgu.05.rcgu.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.std.b489edb20527c47c-cgu.06.rcgu.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.std.b489edb20527c47c-cgu.07.rcgu.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.std.b489edb20527c47c-cgu.08.rcgu.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.std.b489edb20527c47c-cgu.09.rcgu.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.std.b489edb20527c47c-cgu.10.rcgu.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.std.b489edb20527c47c-cgu.11.rcgu.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.std.b489edb20527c47c-cgu.12.rcgu.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.std.b489edb20527c47c-cgu.13.rcgu.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.std.b489edb20527c47c-cgu.14.rcgu.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.std.b489edb20527c47c-cgu.15.rcgu.o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.04mmg6r0xplxffmdeorcz34a2.rcgu.rmeta" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/std-c593c74c43a5f190.76g5bn35ifnknhn176cllb1xr.rcgu.o" "-Wl,--as-needed" "-L" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps" "-L" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/release/deps" "-L" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1/lib/rustlib/aarch64-unknown-nto-qnx700/lib" "-Wl,-Bstatic" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/libpanic_unwind-0167d9a4c20b1c71.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/libobject-e88e046afc3e61fe.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/libmemchr-205f4a9f41af6b4a.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/libaddr2line-92473c4a1f696e22.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/libgimli-1d482bcdd13a105f.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/librustc_demangle-7796bb55423e01e6.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/libstd_detect-c4c882c209895fdb.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/libhashbrown-8e308bdfafb1aa8f.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/librustc_std_workspace_alloc-263094251dee34ab.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/libminiz_oxide-2aff3d763478728c.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/libadler-f2252c605db94e23.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/libunwind-5ef94de33ac336c8.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/libcfg_if-f53f72a5479618c9.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/liblibc-424d3b1142c54799.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/liballoc-3406b7631415168a.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/librustc_std_workspace_core-18b685f88e81f5b7.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/libcore-aed636ac0afd522d.rlib" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/libcompiler_builtins-72a1572b1f6564e4.rlib" "-Wl,-Bdynamic" "-lgcc" "-lc" "-lm" "-lsocket" "-lregex" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1/lib/rustlib/aarch64-unknown-nto-qnx700/lib" "-o" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage1-std/aarch64-unknown-nto-qnx700/release/deps/libstd-c593c74c43a5f190.so" "-shared" "-Wl,-soname=libstd-c593c74c43a5f190.so" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-nodefaultlibs" "-Wl,-z,origin" "-Wl,-rpath,$ORIGIN/../lib"
  = note: /Users/jonathan/qnx700/host/darwin/x86_64/usr/bin/aarch64-unknown-nto-qnx7.0.0-ld: cannot find -lregex
          cc: /Users/jonathan/qnx700/host/darwin/x86_64/usr/bin/aarch64-unknown-nto-qnx7.0.0-ld error 1


error: could not compile `std` (lib) due to 1 previous error
Build completed unsuccessfully in 0:00:09
$ find ~/qnx710 -name "libregex*"
/Users/jonathan/qnx710/target/qnx7/armle-v7/lib/libregex.so
/Users/jonathan/qnx710/target/qnx7/armle-v7/lib/libregex.a
/Users/jonathan/qnx710/target/qnx7/armle-v7/lib/libregex.so.1.sym
/Users/jonathan/qnx710/target/qnx7/armle-v7/lib/libregex.so.1
/Users/jonathan/qnx710/target/qnx7/armle-v7/lib/libregexS.a
/Users/jonathan/qnx710/target/qnx7/aarch64le/lib/libregex.so
/Users/jonathan/qnx710/target/qnx7/aarch64le/lib/libregex.a
/Users/jonathan/qnx710/target/qnx7/aarch64le/lib/libregex.so.1.sym
/Users/jonathan/qnx710/target/qnx7/aarch64le/lib/libregex.so.1
/Users/jonathan/qnx710/target/qnx7/aarch64le/lib/libregexS.a
/Users/jonathan/qnx710/target/qnx7/x86_64/lib/libregex.so
/Users/jonathan/qnx710/target/qnx7/x86_64/lib/libregex.a
/Users/jonathan/qnx710/target/qnx7/x86_64/lib/libregex.so.1.sym
/Users/jonathan/qnx710/target/qnx7/x86_64/lib/libregex.so.1
/Users/jonathan/qnx710/target/qnx7/x86_64/lib/libregexS.a
$ find ~/qnx700 -name "libregex*"
```

Where do I get `libregex` for QNX Neutrino 7.0?

--- a/compiler/rustc_target/src/spec/mod.rs
+++ b/compiler/rustc_target/src/spec/mod.rs
@@ -1837,6 +1837,7 @@ supported_targets! {
 
     ("mips64-openwrt-linux-musl", mips64_openwrt_linux_musl),
 
+    ("aarch64-unknown-nto-qnx700", aarch64_unknown_nto_qnx700),
     ("aarch64-unknown-nto-qnx710", aarch64_unknown_nto_qnx710),
     ("x86_64-pc-nto-qnx710", x86_64_pc_nto_qnx710),
     ("i586-pc-nto-qnx700", i586_pc_nto_qnx700),

--- a/compiler/rustc_target/src/spec/targets/aarch64_unknown_nto_qnx700.rs
+++ b/compiler/rustc_target/src/spec/targets/aarch64_unknown_nto_qnx700.rs
@@ -1,0 +1,38 @@
+use crate::spec::{base, Cc, LinkerFlavor, Lld, Target, TargetOptions};
+
+pub fn target() -> Target {
+    // In QNX, libc does not provide a compatible ABI between versions.
+    // To distinguish between QNX versions, we needed a stable conditional compilation switch,
+    // which is why we needed to implement different targets in the compiler.
+    Target {
+        llvm_target: "aarch64-unknown-unknown".into(),
+        metadata: crate::spec::TargetMetadata {
+            description: Some("ARM64 QNX Neutrino 7.0 RTOS".into()),
+            tier: Some(3),
+            host_tools: Some(false),
+            std: Some(true),
+        },
+        pointer_width: 64,
+        // from: https://llvm.org/docs/LangRef.html#data-layout
+        // e         = little endian
+        // m:e       = ELF mangling: Private symbols get a .L prefix
+        // i8:8:32   = 8-bit-integer, minimum_alignment=8, preferred_alignment=32
+        // i16:16:32 = 16-bit-integer, minimum_alignment=16, preferred_alignment=32
+        // i64:64    = 64-bit-integer, minimum_alignment=64, preferred_alignment=64
+        // i128:128  = 128-bit-integer, minimum_alignment=128, preferred_alignment=128
+        // n32:64    = 32 and 64 are native integer widths; Elements of this set are considered to support most general arithmetic operations efficiently.
+        // S128      = 128 bits are the natural alignment of the stack in bits.
+        data_layout: "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128-Fn32".into(),
+        arch: "aarch64".into(),
+        options: TargetOptions {
+            features: "+v8a".into(),
+            max_atomic_width: Some(128),
+            pre_link_args: TargetOptions::link_args(
+                LinkerFlavor::Gnu(Cc::Yes, Lld::No),
+                &["-Vgcc_ntoaarch64le_cxx"],
+            ),
+            env: "nto70".into(),
+            ..base::nto_qnx::opts()
+        },
+    }
+}

--- a/compiler/rustc_target/src/spec/targets/aarch64_unknown_nto_qnx710.rs
+++ b/compiler/rustc_target/src/spec/targets/aarch64_unknown_nto_qnx710.rs
@@ -1,35 +1,8 @@
-use crate::spec::{base, Cc, LinkerFlavor, Lld, Target, TargetOptions};
+use crate::spec::Target;
 
 pub fn target() -> Target {
-    Target {
-        llvm_target: "aarch64-unknown-unknown".into(),
-        metadata: crate::spec::TargetMetadata {
-            description: Some("ARM64 QNX Neutrino 7.1 RTOS".into()),
-            tier: Some(3),
-            host_tools: Some(false),
-            std: Some(true),
-        },
-        pointer_width: 64,
-        // from: https://llvm.org/docs/LangRef.html#data-layout
-        // e         = little endian
-        // m:e       = ELF mangling: Private symbols get a .L prefix
-        // i8:8:32   = 8-bit-integer, minimum_alignment=8, preferred_alignment=32
-        // i16:16:32 = 16-bit-integer, minimum_alignment=16, preferred_alignment=32
-        // i64:64    = 64-bit-integer, minimum_alignment=64, preferred_alignment=64
-        // i128:128  = 128-bit-integer, minimum_alignment=128, preferred_alignment=128
-        // n32:64    = 32 and 64 are native integer widths; Elements of this set are considered to support most general arithmetic operations efficiently.
-        // S128      = 128 bits are the natural alignment of the stack in bits.
-        data_layout: "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128-Fn32".into(),
-        arch: "aarch64".into(),
-        options: TargetOptions {
-            features: "+v8a".into(),
-            max_atomic_width: Some(128),
-            pre_link_args: TargetOptions::link_args(
-                LinkerFlavor::Gnu(Cc::Yes, Lld::No),
-                &["-Vgcc_ntoaarch64le_cxx"],
-            ),
-            env: "nto71".into(),
-            ..base::nto_qnx::opts()
-        },
-    }
+    let mut base = super::aarch64_unknown_nto_qnx700::target();
+    base.metadata.description = Some("ARM64 QNX Neutrino 7.1 RTOS".into());
+    base.options.env = "nto71".into();
+    base
 }

--- a/library/std/Cargo.toml
+++ b/library/std/Cargo.toml
@@ -35,7 +35,7 @@ miniz_oxide = { version = "0.7.0", optional = true, default-features = false }
 addr2line = { version = "0.22.0", optional = true, default-features = false }
 
 [target.'cfg(not(all(windows, target_env = "msvc")))'.dependencies]
-libc = { version = "0.2.153", default-features = false, features = [
+libc = { version = "0.2.156", default-features = false, features = [
     'rustc-dep-of-std',
 ], public = true }
 

--- a/library/std/src/sys/pal/unix/fs.rs
+++ b/library/std/src/sys/pal/unix/fs.rs
@@ -1717,7 +1717,7 @@ pub fn link(original: &Path, link: &Path) -> io::Result<()> {
     run_path_with_cstr(original, &|original| {
         run_path_with_cstr(link, &|link| {
             cfg_if::cfg_if! {
-                if #[cfg(any(target_os = "vxworks", target_os = "redox", target_os = "android", target_os = "espidf", target_os = "horizon", target_os = "vita"))] {
+                if #[cfg(any(target_os = "vxworks", target_os = "redox", target_os = "android", target_os = "espidf", target_os = "horizon", target_os = "vita", target_os = "nto"))] {
                     // VxWorks, Redox and ESP-IDF lack `linkat`, so use `link` instead. POSIX leaves
                     // it implementation-defined whether `link` follows symlinks, so rely on the
                     // `symlink_hard_link` test in library/std/src/fs/tests.rs to check the behavior.

--- a/library/std/src/sys/pal/unix/process/process_unix.rs
+++ b/library/std/src/sys/pal/unix/process/process_unix.rs
@@ -19,7 +19,8 @@ use crate::sys::process::process_common::*;
 use crate::{fmt, mem, sys};
 
 cfg_if::cfg_if! {
-    if #[cfg(all(target_os = "nto", target_env = "nto71"))] {
+    // This workaround is only needed for QNX 7.0 and 7.1. The bug should have been fixed in 8.0
+    if #[cfg(any(target_env = "nto70", target_env = "nto71"))] {
         use crate::thread;
         use libc::{c_char, posix_spawn_file_actions_t, posix_spawnattr_t};
         use crate::time::Duration;
@@ -189,7 +190,8 @@ impl Command {
     #[cfg(not(any(
         target_os = "watchos",
         target_os = "tvos",
-        all(target_os = "nto", target_env = "nto71"),
+        target_env = "nto70",
+        target_env = "nto71"
     )))]
     unsafe fn do_fork(&mut self) -> Result<pid_t, io::Error> {
         cvt(libc::fork())
@@ -199,7 +201,8 @@ impl Command {
     // or closed a file descriptor while the fork() was occurring".
     // Documentation says "... or try calling fork() again". This is what we do here.
     // See also https://www.qnx.com/developers/docs/7.1/#com.qnx.doc.neutrino.lib_ref/topic/f/fork.html
-    #[cfg(all(target_os = "nto", target_env = "nto71"))]
+    // This workaround is only needed for QNX 7.0 and 7.1. The bug should have been fixed in 8.0
+    #[cfg(any(target_env = "nto70", target_env = "nto71"))]
     unsafe fn do_fork(&mut self) -> Result<pid_t, io::Error> {
         use crate::sys::os::errno;
 
@@ -537,7 +540,7 @@ impl Command {
         // or closed a file descriptor while the posix_spawn() was occurring".
         // Documentation says "... or try calling posix_spawn() again". This is what we do here.
         // See also http://www.qnx.com/developers/docs/7.1/#com.qnx.doc.neutrino.lib_ref/topic/p/posix_spawn.html
-        #[cfg(all(target_os = "nto", target_env = "nto71"))]
+        #[cfg(target_os = "nto")]
         unsafe fn retrying_libc_posix_spawnp(
             pid: *mut pid_t,
             file: *const c_char,

--- a/library/unwind/src/lib.rs
+++ b/library/unwind/src/lib.rs
@@ -165,8 +165,15 @@ extern "C" {}
 extern "C" {}
 
 #[cfg(target_os = "nto")]
-#[link(name = "gcc_s")]
-extern "C" {}
+cfg_if::cfg_if! {
+    if #[cfg(target_env = "nto70")] {
+        #[link(name = "gcc")]
+        extern "C" {}
+    } else {
+        #[link(name = "gcc_s")]
+        extern "C" {}
+    }
+}
 
 #[cfg(target_os = "hurd")]
 #[link(name = "gcc_s")]

--- a/src/doc/rustc/src/platform-support.md
+++ b/src/doc/rustc/src/platform-support.md
@@ -256,6 +256,7 @@ target | std | host | notes
 [`aarch64-kmc-solid_asp3`](platform-support/kmc-solid.md) | ✓ |  | ARM64 SOLID with TOPPERS/ASP3
 [`aarch64-nintendo-switch-freestanding`](platform-support/aarch64-nintendo-switch-freestanding.md) | * |  | ARM64 Nintendo Switch, Horizon
 [`aarch64-unknown-teeos`](platform-support/aarch64-unknown-teeos.md) | ? |  | ARM64 TEEOS |
+[`aarch64-unknown-nto-qnx700`](platform-support/nto-qnx.md) | ? |  | ARM64 QNX Neutrino 7.0 RTOS |
 [`aarch64-unknown-nto-qnx710`](platform-support/nto-qnx.md) | ✓ |  | ARM64 QNX Neutrino 7.1 RTOS |
 `aarch64-unknown-freebsd` | ✓ | ✓ | ARM64 FreeBSD
 [`aarch64-unknown-hermit`](platform-support/hermit.md) | ✓ |  | ARM64 Hermit

--- a/src/doc/rustc/src/platform-support/nto-qnx.md
+++ b/src/doc/rustc/src/platform-support/nto-qnx.md
@@ -24,6 +24,7 @@ Currently, the following QNX Neutrino versions and compilation targets are suppo
 |----------------------|---------------------|:------------:|:----------------:|
 | 7.1 | AArch64 | ✓ | ✓ |
 | 7.1 | x86_64  | ✓ | ✓ |
+| 7.0 | AArch64 | ? | ✓ |
 | 7.0 | x86     |   | ✓ |
 
 Adding other architectures that are supported by QNX Neutrino is possible.
@@ -42,6 +43,23 @@ should be available (in the `$PATH` variable).
 When linking `no_std` applications, they must link against `libc.so` (see example). This is
 required because applications always link against the `crt` library and `crt` depends on `libc.so`.
 This is done automatically when using the standard library.
+
+### Disabling RELocation Read-Only (RELRO)
+
+While not recommended by default, some QNX kernel setups may require the `RELRO` to be disabled with `-C relro_level=off`, e.g. by adding it to the `.cargo/config.toml` file:
+
+```toml
+[target.aarch64-unknown-nto-qnx700]
+rustflags = ["-C", "relro_level=off"]
+```
+
+If your QNX kernel does not allow it, and `relro` is not disabled, running compiled binary would fail with `syntax error: ... unexpected` or similar.  This is due to kernel trying to interpret compiled binary with `/bin/sh`, and obviously failing.  To verify that this is really the case, run your binary with the `DL_DEBUG=all` env var, and look for this output. If you see it, you should disable `relro` as described above.
+
+```text
+Resolution scope for Executable->/bin/sh:
+        Executable->/bin/sh
+        libc.so.4->/usr/lib/ldqnx-64.so.2
+```
 
 ### Small example application
 

--- a/tests/assembly/targets/targets-elf.rs
+++ b/tests/assembly/targets/targets-elf.rs
@@ -54,6 +54,9 @@
 //@ revisions: aarch64_unknown_none_softfloat
 //@ [aarch64_unknown_none_softfloat] compile-flags: --target aarch64-unknown-none-softfloat
 //@ [aarch64_unknown_none_softfloat] needs-llvm-components: aarch64
+//@ revisions: aarch64_unknown_nto_qnx700
+//@ [aarch64_unknown_nto_qnx700] compile-flags: --target aarch64-unknown-nto-qnx700
+//@ [aarch64_unknown_nto_qnx700] needs-llvm-components: aarch64
 //@ revisions: aarch64_unknown_nto_qnx710
 //@ [aarch64_unknown_nto_qnx710] compile-flags: --target aarch64-unknown-nto-qnx710
 //@ [aarch64_unknown_nto_qnx710] needs-llvm-components: aarch64

